### PR TITLE
Fix problems with plugin installation

### LINKED ISSUE #1194: Plugins are broken on 10.3.0-rc1
**Describe the bug**
Plugins won't load from subdirectories of the PluginDir. Trying to upgrade plugins doesn't work properly, as the folder contains the version and overwriting fails. Removed plugins leave behind folders.

**Additional context**
Enough problems that describing them individually is futile. The way we were doing plugins in 10.2. was affected by most of these anyways, this fixes them going forward.

### PR DESCRIPTION
**Pull request by joshuaboniface**
### Changes

Fixes several breaking bugs around the installation, loading, and management of plugins.

1. Restore previous functionality to load plugins from subfolders of the PluginDir, which was potentially broken in #801 (or previously, not sure).
1. Use a directory name based only on the `Name` of the plugin, not the source filename, which contains the version and would be really confusing.
1. Use the new directory name for the deletes if it's present, so that installation and removal happen at that directory level and we don't leave empty folders laying around. Ensures we properly remove additional resources in plugins too, not just the main `.dll` file.
1. Ignore the incoming `target` when installing, and always set it ourself to the proper directory, which would matter when upgrading since the `target` is just the `.dll` filename.
1. Delete an existing target directory before installing if it exists, allowing plugin updates to function properly. Note that not calling any of the plugin removal code is intentional; I suspect that would delete configurations unexpectedly when upgrading which would be annoying. This way, it just replaces the files and then reloads.
1. (Minor) Added some actual debug messages around the plugin download section so failures can be more accurately seen.

The end result is that, for example, the `Anime` plugin will on Debian be installed in `/var/lib/jellyfin/plugins/Anime/` as `Anime.dll`. When removed, the whole `/var/lib/jellyfin/plugins/Anime` directory will be removed. When upgrading to a new version, the directory will be removed automatically as expected and the new plugin installed in its place.

### Issues
Fixes #1194

### CODE REVIEW

**Comment by joshuaboniface** [comment]
@Cvium that's what this fixes.

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 1]
```diff
@@ -1047,7 +1047,7 @@ private void SetStaticProperties()
         private async void PluginInstalled(object sender, GenericEventArgs<PackageVersionInfo> args)
         {
             string dir = Path.Combine(ApplicationPaths.PluginsPath, Path.GetFileNameWithoutExtension(args.Argument.targetFilename));
-            var types = Directory.EnumerateFiles(dir, "*.dll", SearchOption.TopDirectoryOnly)
+            var types = Directory.EnumerateFiles(dir, "*.dll", SearchOption.AllDirectories)
```
Comment (cvium) [inline Emby.Server.Implementations/ApplicationHost.cs]: I don't think this change is necessary

Comment (joshuaboniface) [reply Emby.Server.Implementations/ApplicationHost.cs]: Yes, this one is iffy. With or without this change no plugins seemed to load without a restart, so zi couldn't verify how required it was. I changed it so they're both the same.

**Comment by joshuaboniface** [comment]
Sorry if my original message isn't clear, I worded that very oddly.

Before #801, as far as I remember, loading from subfolders worked. But this search `TopDirectoryOnly` prevents it from searching those subfolders. So I changed to `AllDirectories` to have it search subdirectories too. I didn't blame past there, so I'm not sure what it was before, but it seems to work fine.

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 2]
```diff
@@ -85,9 +85,11 @@ private void OnPluginUpdated(IPlugin plugin, PackageVersionInfo newVersion)
         private void OnPluginInstalled(PackageVersionInfo package)
         {
             _logger.LogInformation("New plugin installed: {0} {1} {2}", package.name, package.versionStr ?? string.Empty, package.classification);
+            _logger.LogDebug("{String}", package.name);
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]: Seems unnecessary
```suggestion
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 3]
```diff
@@ -85,9 +85,11 @@ private void OnPluginUpdated(IPlugin plugin, PackageVersionInfo newVersion)
...
 
             PluginInstalled?.Invoke(this, new GenericEventArgs<PackageVersionInfo> { Argument = package });
 
+            _logger.LogDebug("{String}", package.name);
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]: Seems unnecessary
```suggestion
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 4]
```diff
@@ -518,12 +520,12 @@ private async Task PerformPackageInstallation(IProgress<double> progress, string
                 return;
             }
 
-            if (target == null)
-            {
-                target = Path.Combine(_appPaths.PluginsPath, Path.GetFileNameWithoutExtension(package.targetFilename));
-            }
+            // Always override the passed-in target (which is a file) and figure it out again
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]: We should change the function signature then...

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 5]
```diff
@@ -574,7 +585,18 @@ public void UninstallPlugin(IPlugin plugin)
             _applicationHost.RemovePlugin(plugin);
 
             var path = plugin.AssemblyFilePath;
-            _logger.LogInformation("Deleting plugin file {0}", path);
+            bool is_path_directory = false;
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]:
```suggestion
            bool isDirectory = false;
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 6]
```diff
@@ -574,7 +585,18 @@ public void UninstallPlugin(IPlugin plugin)
...
+            // Check if we have a plugin directory we should remove too
+            if (Path.GetDirectoryName(plugin.AssemblyFilePath) != _appPaths.PluginsPath)
+            {
+                path = Path.GetDirectoryName(plugin.AssemblyFilePath);
+                is_path_directory = true;
+                _logger.LogInformation("Deleting plugin directory {0}", path);
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]:
```suggestion
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 7]
```diff
@@ -585,7 +607,14 @@ public void UninstallPlugin(IPlugin plugin)
                 path = file;
             }
 
-            _fileSystem.DeleteFile(path);
+            if (is_path_directory)
+            {
+                Directory.Delete(path, true);
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]:
```suggestion
                _logger.LogInformation("Deleting plugin directory {Path}", path);
                Directory.Delete(path, true);
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 8]
```diff
@@ -574,7 +585,18 @@ public void UninstallPlugin(IPlugin plugin)
...
-            _logger.LogInformation("Deleting plugin file {0}", path);
+            bool is_path_directory = false;
+            // Check if we have a plugin directory we should remove too
+            if (Path.GetDirectoryName(plugin.AssemblyFilePath) != _appPaths.PluginsPath)
+            {
+                path = Path.GetDirectoryName(plugin.AssemblyFilePath);
+                is_path_directory = true;
+                _logger.LogInformation("Deleting plugin directory {0}", path);
+            }
+            else
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]:
```suggestion
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 9]
```diff
@@ -574,7 +585,18 @@ public void UninstallPlugin(IPlugin plugin)
...
+            {
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]:
```suggestion
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 10]
```diff
@@ -574,7 +585,18 @@ public void UninstallPlugin(IPlugin plugin)
...
+                _logger.LogInformation("Deleting plugin file {0}", path);
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]:
```suggestion
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 11]
```diff
@@ -574,7 +585,18 @@ public void UninstallPlugin(IPlugin plugin)
...
+            }
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]:
```suggestion
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 12]
```diff
@@ -585,7 +607,14 @@ public void UninstallPlugin(IPlugin plugin)
...
             }
 
-            _fileSystem.DeleteFile(path);
+            if (is_path_directory)
+            {
+                Directory.Delete(path, true);
+            }
+            else
+            {
+                _fileSystem.DeleteFile(path);
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]:
```suggestion
                 _logger.LogInformation("Deleting plugin file {Path}", path);
                _fileSystem.DeleteFile(path);
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 13]
```diff
@@ -574,7 +585,18 @@ public void UninstallPlugin(IPlugin plugin)
             _applicationHost.RemovePlugin(plugin);
 
             var path = plugin.AssemblyFilePath;
-            _logger.LogInformation("Deleting plugin file {0}", path);
+            bool is_path_directory = false;
+            // Check if we have a plugin directory we should remove too
+            if (Path.GetDirectoryName(plugin.AssemblyFilePath) != _appPaths.PluginsPath)
+            {
+                path = Path.GetDirectoryName(plugin.AssemblyFilePath);
+                is_path_directory = true;
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]:
```suggestion
                isDirectory = true;
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs [hunk 14]
```diff
@@ -585,7 +607,14 @@ public void UninstallPlugin(IPlugin plugin)
                 path = file;
             }
 
-            _fileSystem.DeleteFile(path);
+            if (is_path_directory)
```
Comment (cvium) [inline Emby.Server.Implementations/Updates/InstallationManager.cs]:
```suggestion
            if (isDirectory)
```

File: Emby.Server.Implementations/Updates/InstallationManager.cs (hunk 4 above)
Comment (joshuaboniface) [reply Emby.Server.Implementations/Updates/InstallationManager.cs]: That's far out of my element, I went for quick-and-dirty right now.

**Comment by joshuaboniface** [comment]
@cvium All addressed :+1:

File: Emby.Server.Implementations/Updates/InstallationManager.cs (hunk 4 above)
Comment (cvium) [reply Emby.Server.Implementations/Updates/InstallationManager.cs]: Could you add a TODO then?

Comment (joshuaboniface) [reply Emby.Server.Implementations/Updates/InstallationManager.cs]: Done!

**Comment by faust93** [comment]
Plugins enumeration and loading seems fixed with those patches. Relevant control elements for every plugin are also in place on the admin's dashboard. But there're actually no plugin entries on the home page, either web or mobile app. For example IPTV installed and configured but it's actually impossible to use it.

**Comment by joshuaboniface** [comment]
@faust93 Do you mean in the Plugins page? All plugin configuration is done from there and it should list all your installed plugins.

**Comment by faust93** [comment]
@joshuaboniface Plugins configuration items are there, yes. I'm able to configure IPTV plugin for ex. But plugin does not appear on the 'Home' page, in 2.2.0 version I got it listed under 'My Media' heading usually.

**Comment by joshuaboniface** [comment]
OK that sounds like a more involved issue, perhaps with the IPTV plugin specifically. Are you able to create an issue (with screenshots preferably) of what you're seeing here @faust93?

**Pull request merged**